File: backend/services/demucs_service.py
"""
services/demucs_service.py — Vocal isolation using Demucs (by Facebook Research).

We chose Demucs over Spleeter because:
  • Spleeter requires old TensorFlow 2.x which conflicts on modern Windows/Python.
  • Demucs (htdemucs) produces significantly higher-quality vocal separation.
  • Demucs installs cleanly via pip on Python 3.10+.

The heavy lifting runs in a subprocess so it doesn't block the event loop.
"""

import os
import sys
import subprocess
import logging
from pathlib import Path

logger = logging.getLogger(__name__)

# Demucs model to use — can be overridden via .env
DEMUCS_MODEL = os.getenv("DEMUCS_MODEL", "htdemucs")


def isolate_vocals(
    audio_path: str,
    output_dir: str,
    progress_callback=None,
) -> str:
    """
    Run Demucs vocal isolation on *audio_path*.

    Parameters
    ----------
    audio_path : str
        Absolute path to the input audio file (MP3, WAV, FLAC, …).
    output_dir : str
        Directory where Demucs will write its stems subdirectory.
    progress_callback : callable, optional
        Called with a float (0–100) whenever we have a progress estimate.

    Returns
    -------
    str
        Path to the extracted vocals WAV file.

    Raises
    ------
    RuntimeError
        If Demucs exits with a non-zero return code.
    """
    audio_path = str(Path(audio_path).resolve())
    output_dir = str(Path(output_dir).resolve())

    logger.info(f"[Demucs] Starting vocal isolation for: {audio_path}")
    if progress_callback:
        progress_callback(2)  # starting demucs

    # Build the subprocess command.
    # --two-stems vocals  → only produce vocals + no_vocals (faster than all 4 stems)
    # -n NAME             → model name (correct flag in demucs 4.x; NOT --model)
    # -o output_dir       → where to save output
    cmd = [
        sys.executable, "-m", "demucs",
        "--two-stems", "vocals",
        "-n", DEMUCS_MODEL,
        "-o", output_dir,
        audio_path,
    ]

    logger.debug(f"Demucs command: {' '.join(cmd)}")

    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=1800,  # 30-minute timeout (large files on CPU can be slow)
        )
    except subprocess.TimeoutExpired:
        raise RuntimeError("Demucs timed out after 30 minutes — try a shorter audio file.")

    if result.returncode != 0:
        raise RuntimeError(
            f"Demucs failed (exit {result.returncode}):\n{result.stderr}"
        )

    if progress_callback:
        progress_callback(100)

    # Demucs writes to: output_dir/<model>/<stem_name>/vocals.wav
    stem_name = Path(audio_path).stem
    vocals_path = Path(output_dir) / DEMUCS_MODEL / stem_name / "vocals.wav"

    if not vocals_path.exists():
        # Fallback: search for any vocals.wav under output_dir
        candidates = list(Path(output_dir).rglob("vocals.wav"))
        if candidates:
            vocals_path = candidates[0]
        else:
            raise RuntimeError(
                f"Demucs finished but vocals.wav not found under {output_dir}.\n"
                f"Demucs stdout: {result.stdout}"
            )

    logger.info(f"[Demucs] Vocals isolated: {vocals_path}")
    return str(vocals_path)

File: backend/src/main/java/com/lyrictranslate/service/DemucsService.java
package com.lyrictranslate.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.logging.Logger;

/**
 * DemucsService — Vocal isolation via Facebook's Demucs model.
 *
 * Direct Java equivalent of services/demucs_service.py::isolate_vocals().
 * Runs Demucs as a Python subprocess via ProcessBuilder.
 *
 * ProcessBuilder replaces Python's subprocess.run().
 */
@Service
public class DemucsService {

    private static final Logger log = Logger.getLogger(DemucsService.class.getName());

    @Value("${app.demucs-model:htdemucs}")
    private String demucsModel;

    /**
     * Run Demucs vocal isolation on audioPath.
     *
     * @param audioPath  Absolute path to the input audio file
     * @param outputDir  Directory where Demucs will write its stems
     * @param progressCallback  Runnable called with estimated progress (not used in subprocess mode)
     * @return Path to the extracted vocals WAV file
     */
    public String isolateVocals(String audioPath, String outputDir) throws Exception {
        audioPath = Paths.get(audioPath).toAbsolutePath().toString();
        outputDir = Paths.get(outputDir).toAbsolutePath().toString();

        log.info("[Demucs] Starting vocal isolation for: " + audioPath);

        // Find Python executable — uses the same venv Python that has demucs installed
        String pythonExe = findPython();

        List<String> cmd = List.of(
            pythonExe, "-m", "demucs",
            "--two-stems", "vocals",
            "-n", demucsModel,
            "-o", outputDir,
            audioPath
        );

        log.info("[Demucs] Command: " + String.join(" ", cmd));

        ProcessBuilder pb = new ProcessBuilder(cmd);
        pb.redirectErrorStream(true);
        pb.directory(new File(System.getProperty("user.dir")));

        Process process = pb.start();
        String output = new String(process.getInputStream().readAllBytes());

        int exitCode = process.waitFor();
        if (exitCode != 0) {
            throw new RuntimeException("Demucs failed (exit " + exitCode + "):\n" + output);
        }

        log.info("[Demucs] Finished successfully");

        // Demucs writes to: outputDir/<model>/<stem_name>/vocals.wav
        String stemName  = Paths.get(audioPath).getFileName().toString()
                               .replaceAll("\\.[^.]+$", "");  // strip extension
        Path vocalsPath  = Paths.get(outputDir, demucsModel, stemName, "vocals.wav");

        if (!Files.exists(vocalsPath)) {
            // Fallback: search recursively for any vocals.wav
            Optional<Path> found = Files.walk(Paths.get(outputDir))
                .filter(p -> p.getFileName().toString().equals("vocals.wav"))
                .findFirst();
            if (found.isPresent()) {
                vocalsPath = found.get();
            } else {
                throw new RuntimeException(
                    "Demucs finished but vocals.wav not found under " + outputDir
                );
            }
        }

        log.info("[Demucs] Vocals isolated: " + vocalsPath);
        return vocalsPath.toString();
    }

    /**
     * Find the Python executable — prefers the venv Python inside backend/.
     * Falls back to system python3 / python.
     */
    public static String findPython() {
        // Check for venv alongside the backend directory
        String[] candidates = {
            "venv/Scripts/python.exe",    // Windows venv
            "venv/bin/python",            // Linux/Mac venv
            "python3",
            "python"
        };
        for (String candidate : candidates) {
            File f = new File(candidate);
            if (f.exists()) return f.getAbsolutePath();
        }
        return "python";  // last resort — relies on PATH
    }
}
